src/primitive.ts

```diff
@@ -126,51 +126,66 @@ export function parseString (str: string, ptr = 0, endPtr = str.length): string
 	return parsed + str.slice(sliceStart, endPtr - 1)
 }
 
-export function parseValue (value: string, toml: string, ptr: number): boolean | number | TomlDate {
+export enum IntegerParsing { NUMBER_OR_ERROR, NUMBER_OR_BIGINT, BIGINT_ONLY }
+
+export function parseValue(value: string, toml: string, ptr: number, integerParsing: IntegerParsing = IntegerParsing.NUMBER_OR_ERROR): boolean | number | bigint | TomlDate {
 	// Constant values
-	if (value === 'true') return true
-	if (value === 'false') return false
-	if (value === '-inf') return -Infinity
-	if (value === 'inf' || value === '+inf') return Infinity
-	if (value === 'nan' || value === '+nan' || value === '-nan') return NaN
+	if (value === 'true') return true;
+	if (value === 'false') return false;
+	if (value === '-inf') return -Infinity;
+	if (value === 'inf' || value === '+inf') return Infinity;
+	if (value === 'nan' || value === '+nan' || value === '-nan') return NaN;
 
-	if (value === '-0') return 0 // Avoid FP representation of -0
+	if (value === '-0') return 0; // Avoid FP representation of -0
 
 	// Numbers
-	let isInt
-	if ((isInt = INT_REGEX.test(value)) || FLOAT_REGEX.test(value)) {
+	const isInt = INT_REGEX.test(value);
+	const isFloat = FLOAT_REGEX.test(value);
+	if (isInt || isFloat) {
 		if (LEADING_ZERO.test(value)) {
 			throw new TomlError('leading zeroes are not allowed', {
 				toml: toml,
 				ptr: ptr
-			})
+			});
 		}
 
-		let numeric = +(value.replace(/_/g, ''))
```

Comment:
The only special handling comes from integer values. So I think quite a lot can be simplified.

```ts
value = value.replace(...)
let numeric = +value
if (isNaN(value)) ...

if (isInt) {
  // isInt is no longer used, so we can reuse it to keep whether it's a safe int or not.
  if (!(isInt = Number.isSafeInteger(numeric)) && !useBigIntegers) ... // Precision loss.
  if (!isInt || useBigIntegers == 1) numeric = BigInt(value)
}
```

Reply:
We don't even need to reuse the variable, actually (I do frown upon that)

Reply:
Yeah I'd frown upon that in most other situations too. Again, here I'm in the shoes of someone optimizing for code size, and sometimes sacrifices needs to be made. It's not *too* bad either, as the type remains the same at least. 🤷🏻‍♀️

I'm sure you've seen worse in the code already, I used some very cursed tricks here and there 😅

Reply:
TODO : I think that version doesn't support numbers over e310, since once converted to number they are nan but yet they may be integers supported by bigint.

Reply:
I meant higher than that value, they can be written as a string of 300+ digits

Reply:
Are you sure they get converted to NaN? I get Infinity after a quick test, which is fine (it's not NaN, and it's not a safe integer either) 🤔 

```
Node REPL
> +('9'.repeat(310))
Infinity
```

Reply:
I added tests for that.